File: algorithm/baekjoon/src/main/java/com/keunwon/algorithm/bruteforce/Problem15650.java
package com.keunwon.algorithm.bruteforce;

import java.util.Scanner;

public class Problem15650 {

    public static int[] arr;
    public static boolean[] visited;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        final int n = sc.nextInt();
        final int m = sc.nextInt();

        visited = new boolean[n];
        arr = new int[m];

        dfs(n, m, 0);
    }

    public static void dfs(int n, int m, int depth) {
        if (m == depth) {
            for (int i : arr) {
                System.out.print(i + " ");
            }
            System.out.println();
            return;
        }

        for (int i = 0; i < n; i++) {
            if (visited[i]) { continue; }

            visited[i] = true;
            arr[depth] = i + 1;
            dfs(n, m, depth + 1);

            for (int j = i + 1; j < n; j++) {
                visited[j] = false;
            }
        }
    }
}

File: algorithm/kotlin-algorithm/src/main/kotlin/com/keunwon/algorithm/baekjoon/Problem15650.kt
package com.keunwon.algorithm.baekjoon

/**
 * Title: N과 M (2)
 * Level: 실버-3
 **/
class Problem15650 {
    private val answer = mutableListOf<String>()

    fun solution(n: Int, m: Int): List<String> {
        dfs(0, 1, n, IntArray(m), BooleanArray(n + 1))
        return answer.sorted()
    }

    private fun dfs(depth: Int, startIndex: Int, n: Int, picks: IntArray, visited: BooleanArray) {
        if (depth == picks.size) {
            answer.add(picks.joinToString(" "))
            return
        }

        for (i in startIndex..n) {
            if (visited[i]) continue

            visited[i] = true
            picks[depth] = i
            dfs(depth + 1, i + 1, n, picks, visited)
            visited[i] = false
        }
    }
}

fun main() {
    val (n, m) = readLine()!!.split(" ").map { it.toInt() }
    Problem15650().solution(n, m).forEach(::println)
}
